### PR TITLE
[core] Clean `filtering.DataGrid.test.tsx`

### LINKED ISSUE #3518: [DataGrid] Filtering tests cleanup
### Duplicates

- [X] I have searched the existing issues

### Current behavior 😯

During the last retro, we identified a potential improvement of the filtering tests. The current state they are in right now is hard to navigate and there aren't logically separated. It makes sense to invest some time into cleaning up these tests as filtering is one of the more complex features we have. We can think of also splitting the tests into 2 parts - one for filtering being done through the filter panel and one using the API directly.

### Expected behavior 🤔

Filtering tests should be logically organized and easier to navigate and maintain.

### PR DESCRIPTION
**Pull request by flaviendelangle**
Closes #3518

- [x] Unify all the column type tests with a `getRows` which allow to to cover more scenarios for each operator (they replace the `forEach` we had for some column types)

```ts
it('should filter with operator "="', () => {
  expect(getRows({ operatorValue: '=', value: 1974 })).to.deep.equal(['1974']);
  expect(getRows({ operatorValue: '=', value: 0 })).to.deep.equal(['', '0']);
  expect(getRows({ operatorValue: '=', value: undefined })).to.deep.equal(ALL_ROWS);
  expect(getRows({ operatorValue: '=', value: '' })).to.deep.equal(ALL_ROWS);
});
```

- [x] Always test the nullish values for `operatorValue`

- [x] Handle null value in `singleSelect` (they were causing crashes)

- [x] Remove duplicated tests (sometimes we have two times the same tests for basic use-cases, especially in the `string` column type.

- [x] Add `TODO` for filtering behavior that I find questionable

- [x] Move filter panel tests to a standalone file (only copy paste, I did not touche those tests)

### CODE REVIEW

File: packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 1]
```diff
@@ -179,1517 +56,1080 @@ describe('<DataGrid /> - Filter', () => {
...
+              Toolbar: GridToolbar,
+            }}
+          />
+        </div>
       );
+    }).not.to.throw();
+  });
 
-      expect(getColumnValues(0)).to.deep.equal(['Adidas', 'Puma']);
+  it('should apply the valueParser onto the filter value', () => {
```
Comment (alexfauquette) [inline packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:1086]: It sounds like a duplicate of 

```it('should support `valueParser`', () => {```

File: packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 2]
```diff
@@ -179,1517 +56,1080 @@ describe('<DataGrid /> - Filter', () => {
...
+    );
+    expect(getColumnValues(0)).to.deep.equal(['0.5']);
+  });
 
-      fireEvent.change(getByText(screen.getByRole('tooltip'), 'true').parentNode, {
-        target: { value: 'false' },
-      });
-      expect(getColumnValues(0)).to.deep.equal(['Nike']);
-    });
+  it('should include value-less operators when displaying the number of active filters', () => {
```
Comment (alexfauquette) [inline packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:1120]: Maybe should add the opposite test: "should not consider incomplete operators when displaying the number of active filters"

**Review by alexfauquette — APPROVED** [review]

File: packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 3]
```diff
@@ -179,1517 +56,1080 @@ describe('<DataGrid /> - Filter', () => {
...
-              },
-            ]}
-          />,
-        );
-        expect(getColumnValues()).to.deep.equal(['12/1/2000', '1/1/2001', '1/1/2002']);
-      });
+    it('should filter with operator "isAnyOf"', () => {
+      expect(getRows({ operatorValue: 'isAnyOf', value: ['France (fr)'] })).to.deep.equal([
+        'France (fr)',
+      ]);
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:328]: Can't be bad to add it

File: packages/grid/x-data-grid/src/tests/filterPanel.DataGrid.test.tsx [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    expect(getColumnValues(0)).to.deep.equal(['Puma']);
+
+    setColumnValue('slogan');
+
+    expect(getColumnValues(0)).to.deep.equal([]);
+    expect(screen.getByRole('combobox', { name: 'Operators' }).value).to.equal('equals');
+    expect(screen.getByRole('textbox', { name: 'Value' }).value).to.equal('Puma');
+  });
+
+  it('should reset value if operator is not available for the new column', () => {
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/grid/x-data-grid/src/tests/filterPanel.DataGrid.test.tsx:183]: Yes, I propose we do two follow up PRs for `filterPanel.DataGrid.test.tsx` and `filtering.DataGridPro.test.tsx` :+1:

File: packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 5]
```diff
@@ -179,1517 +56,1080 @@ describe('<DataGrid /> - Filter', () => {
...
+            { field: 'id' },
+            {
+              field: 'status',
+              type: 'singleSelect',
+              valueOptions,
+              valueParser: (value) => {
+                if (typeof value === 'number') {
+                  return valueOptions.find((option) => option.value === value);
+                }
+                return value;
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:986]: Not sure to understand your point here

[the remark this answers is not in the document]
Comment (alexfauquette) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:991]: Here is the code for filtering single select columns. They all use `parseObjectValue` such that all the following combinations works
|`filterItem.value`| `value`|
|---|---|
| `42` | `42` |
| `42` | `{value: 42, label:"toto"}` |
| `{value: 42, label:"toto"}` | `42` |
| `{value: 42, label:"toto"}` | `{value: 42, label:"toto"}` |

If I correctly understand, the `valueParser` transforms `filterItem.value=42` to `filterItem.value={value: 42, label:"toto"}`
But that does not change anything in the filtering behavior. So if I remove the line where `valueParser` is called, this test will not fail.

You could do something like going from `filterItem.value='toto'` to `filterItem.value=42`. In such a case, filtering behavior will be modified by the `valueParser`
```js
const parseObjectValue = (value) =>
  typeof value === 'object' ? (value as { value: any; label: string }).value : value;


export const getGridSingleSelectOperators: () => GridFilterOperator[] = () => [
  {
    value: 'is',
    getApplyFilterFn: (filterItem: GridFilterItem) => {
      if (filterItem.value == null || filterItem.value === '') {
        return null;
      }
      return ({ value }): boolean => parseObjectValue(value) === parseObjectValue(filterItem.value);
    },
    InputComponent: GridFilterInputSingleSelect,
  },
```

[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx]: Done :+1:

**Review by alexfauquette — APPROVED** [review]
That's much clearer with this two distinct file and the getRows helper :)

**Pull request merged**